### ISSUE TITLE
buildmaster: tries to rebuild an already built recipe

### ISSUE THREAD

**Issue by korli** [haikuports/haikuporter#146]
From: http://vmpkg.haiku-os.org/master/x86_64/?buildrunDir=571&viewMode=expanded

The previous run had built libopenshot_audio successfully: http://vmpkg.haiku-os.org/master/x86_64/?buildrunDir=570&viewMode=expanded  based on this commit https://github.com/haikuports/haikuports/commit/6ec6338f0538b1466e0b2db8269ce877c63a5a4a

The buildrun 571 is based on commit https://github.com/haikuports/haikuports/commit/e150defa57283dd4ef3a7065c6aa02f543fa4931 

The recipes affected by buildrun 571 should be openshot and libopenshot. Somehow haikuporter thinks it needs to rebuild libopenshot-audio.

maybe because libopenshot-audio starts with libopenshot?

**Comment by korli** [comment]
@mmlr could it be that the list of ports changed was wrong? https://github.com/haikuports/haikuporter/blob/master/buildmaster/bin/buildmaster#L47

**Comment by mmlr** [comment]
@korli no, the list of ports was correct, i.e. libopenshot-0.1.9 and openshot-2.4.1.

For reference:

```
moving from 6ec6338f0538b1466e0b2db8269ce877c63a5a4a to e150defa57283dd4ef3a7065
c6aa02f543fa4931
Checking if any package-infos need to be updated ...
Looking for stale dependency-infos ...
ports to be built: libopenshot-0.1.9
openshot-2.4.1
Warning: skipped port libopenshot_audio-0.1.5-1: some packages already exist, revision bump required
Local builders count: 0
Checking if any package-infos need to be updated ...
Looking for stale dependency-infos ...
======================================================================
media-libs::libopenshot-0.1.9
======================================================================
Checking if package dependencies need to be updated ...
        updating dependency infos of libopenshot-0.1.9
The following built dependencies were found:
======================================================================
media-video::openshot-2.4.1
======================================================================
Checking if package dependencies need to be updated ...
        updating dependency infos of openshot-2.4.1
The following built dependencies were found:
        /builds/buildmaster/master/x86_64/haikuports/repository/libopenshot_audio-0.1.5.DependencyInfo
        /builds/buildmaster/master/x86_64/haikuports/repository/libopenshot-0.1.9.DependencyInfo
        /builds/buildmaster/master/x86_64/haikuports/repository/libopenshot_devel-0.1.9.DependencyInfo
The following required ports will be built first:
        media-libs::libopenshot_audio-0.1.5
        media-libs::libopenshot-0.1.9
======================================================================
media-libs::libopenshot_audio-0.1.5
======================================================================
Checking if package dependencies need to be updated ...
        updating dependency infos of libopenshot_audio-0.1.5
The following built dependencies were found:
```

So it was openshot that pulled in both libopenshot and libopenshot_audio. I do not have an explanation for that. Since openshot does not directly depend on libopenshot_audio I guess it was pulled in via libopenshot. But it can't be as fundamental as "everything down from any built dependency is built as well" as that would have pulled in a lot more indirect dependencies. I'll try to reproduce the case and add more debug output.

**Comment by mmlr** [comment]
Ok, the issue was quite a bit larger and actually comprised of two different bugs coming together.

The reason that libopenshot_audio was pulled in is actually that a HPKG of an older version of libopenshot (0.1.8) was present (as libopenshot 0.1.9 was only to be built with that run) and it required lib:libopenshot_audio >= 5. This isn't provided by the current libopenshot_audio-0.1.5 at all, as it has a compat >= 6 declared. Due to a bug, these compat declarations from recipes were completely ignored though. This first part of the problem was fixed in ddd2e5fe71a07378b6ba1f5967996db166c89d48.

With the compat declarations fixed, it would now simply fail to build due to the unmet dependency of the outdated libopenshot-0.1.8. The dependency resolution never progressed to the valid solution of building libopenshot-0.1.9 which is compatible with the already built libopenshot_audio-0.1.5. This part was fixed in c5705a423037cfa636a978fe00f56fdf41266607 which would now cause the unresolved libopenshot-0.1.8 to be removed from dependency resolution and it being restarted to find the valid solution.

**Comment by korli** [comment]
Thanks for the fix and explanations. The case doesn't happen often for sure.
